numpyro/distributions/util.py

```diff
@@ -346,6 +346,69 @@ def clamp_probs(probs):
     return jnp.clip(probs, a_min=finfo.tiny, a_max=1. - finfo.eps)
 
 
+def von_mises_centered(key, concentration, shape=(), dtype=jnp.float64):
+    return _von_mises_centered(key, concentration, shape, dtype)
+
+
+@partial(jit, static_argnums=(2, 3))
+def _von_mises_centered(key, concentration, shape, dtype):
+    """ Compute centered von Mises samples using rejection sampling from [1] with wrapped Cauchy proposal.
+
+        *** References ***
+        [1] Luc Devroye "Non-Uniform Random Variate Generation", Springer-Verlag, 1986;
+            Chapter 9, p. 473-476. http://www.nrbook.com/devroye/Devroye_files/chapter_nine.pdf
+
+        :param sample_shape: shape of samples
+        :param rng_key: random number generator key
+        :return: centered samples from von Mises
+    """
+
+    # Cutoff from TensorFlow probability
+    # (https://github.com/tensorflow/probability/blob/f051e03dd3cc847d31061803c2b31c564562a993/tensorflow_probability/python/distributions/von_mises.py#L567-L570)
+    s_cutoff_map = {jnp.dtype(jnp.float16): 1.8e-1,
+                    jnp.dtype(jnp.float32): 2e-2,
+                    jnp.dtype(jnp.float64): 1.2e-4}
+
+    r = 1. + jnp.sqrt(1. + 4. * concentration ** 2)
+    rho = (r - jnp.sqrt(2. * r)) / (2. * concentration)
+    s_exact = (1. + rho ** 2) / (2. * rho)
+
+    s_approximate = 1. / concentration
+
+    s_cutoff = s_cutoff_map.get(dtype)
+
+    s = jnp.where(concentration > s_cutoff, s_exact, s_approximate)
+
+    def body_function(i, *args):
+        done, _, w = args[0]
+        nonlocal key
+        uni_ukey, uni_vkey, key = random.split(key, 3)
+
+        u = random.uniform(key=uni_ukey, shape=shape, dtype=concentration.dtype, minval=-1., maxval=1.)
+        z = jnp.cos(jnp.pi * u)
+        w = jnp.where(done, w, (1. + s * z) / (s + z))  # Update where not done
+
+        y = concentration * (s - w)
+        v = random.uniform(key=uni_vkey, shape=shape, dtype=concentration.dtype, minval=-1., maxval=1.)
+
+        accept = (y * (2. - y) >= v) | (jnp.log(y / v) + 1. >= y)
+
+        return accept | done, u, w
+
+    init_done = jnp.zeros(shape, dtype=bool)
+    init_u = jnp.zeros(shape)
+    init_w = jnp.zeros(shape)
+
+    done, u, w = lax.fori_loop(
```

Comment:
I just test, the following code works and generates correct distribution. Otherwise, using the constant global `key` for all steps will generate incorrect distribution.

```python
    def cond_function(val):
        i, _, done, _, _ = val
        return (i < 100) & (~jnp.all(done))

    def body_function(val):
        i, key, done, _, w = val
        uni_ukey, uni_vkey, key = random.split(key, 3)

        u = random.uniform(key=uni_ukey, shape=shape, dtype=dtype, minval=-1., maxval=1.)
        z = jnp.cos(jnp.pi * u)
        w = jnp.where(done, w, (1. + s * z) / (s + z))  # Update where not done

        y = concentration * (s - w)
        v = random.uniform(key=uni_vkey, shape=shape, dtype=dtype, minval=-1., maxval=1.)

        accept = (y * (2. - y) >= v) | (jnp.log(y / v) + 1. >= y)

        return i + 1, key, accept | done, u, w

    init_done = jnp.zeros(shape, dtype=bool)
    init_u = jnp.zeros(shape)
    init_w = jnp.zeros(shape)

    _, _, _, u, w = lax.while_loop(
        cond_fun=cond_function,
        body_fun=body_function,
        init_val=(jnp.array(0), key, init_done, init_u, init_w)
    )
```

Reply:
Thanks for checking! It got late, so I didn't start on the for loop. I'll make the necessary changes.

Reply:
Your solution to the `while_loop` fixed the samples! Thanks a lot for all the great feedback :D